### PR TITLE
dpg, expose header and query param to method signature

### PR DESCRIPTION
**Pull request by haolingdong-msft**
fix https://github.com/Azure/autorest.java/issues/1457

1. expose header and query parameters to method signature.
2. remove required query and header parameters in Java doc.
3. support different types as query and header parameters and add tests for it. Tested types: int, boolean, array as CSV, array as multi (explode), array as CSV/multi with item in integer, date-time, duration

Please take a look to see if the generated client method and proxy method look fine. @srnagar 

Generated client method like this: https://github.com/Azure/autorest.java/blob/2434674d50c45b60126b7f0f08897a28a04738bf/protocol-resilience-test/llcupdate1/src/main/java/fixtures/llcresi/DpgClient.java#L92

Generated proxy method like this:
https://github.com/Azure/autorest.java/blob/2434674d50c45b60126b7f0f08897a28a04738bf/protocol-tests/src/main/java/fixtures/llcupdate1/implementation/ParamsImpl.java#L91-L96

### CODE REVIEW

**Comment by weidongxu-microsoft** [comment]
And these query/header would not appear in Javadoc.

File: protocol-resilience-test/llcinitial/src/main/java/fixtures/llcresi/DpgClient.java [hunk 1]
```diff
@@ -61,7 +61,6 @@ public Response<BinaryData> headNoParamsWithResponse(RequestOptions requestOptio
      * <table border="1">
      *     <caption>Query Parameters</caption>
      *     <tr><th>Name</th><th>Type</th><th>Required</th><th>Description</th></tr>
-     *     <tr><td>parameter</td><td>String</td><td>Yes</td><td>I am a required parameter</td></tr>
      * </table>
```
Comment (haolingdong-msft) [inline protocol-resilience-test/llcinitial/src/main/java/fixtures/llcresi/DpgClient.java:64]: Here is what generated code look like

**Comment by weidongxu-microsoft** [comment]
Consider adding a swagger with required query/header with type int, boolean, array as CSV, array as multi (explode).

Also verify "accept" header in ProxyMethod can be overwritten with `RequestOptions.setHeader`.

File: partial-update-tests/generated/src/main/java/fixtures/bodystring/implementation/EnumsImpl.java [hunk 2]
```diff
@@ -114,6 +122,7 @@ Mono<Response<String>> getReferenced(
         Mono<Response<Void>> putReferenced(
                 @HostParam("$host") String host,
                 @BodyParam("application/json") BinaryData enumStringBody,
+                @HeaderParam("Accept") String accept,
```
Comment (srnagar) [inline partial-update-tests/generated/src/main/java/fixtures/bodystring/implementation/EnumsImpl.java:125]: Should we add headers, query params before body?

**Review by srnagar — APPROVED** [review]

File: partial-update-tests/generated/src/main/java/fixtures/bodystring/implementation/EnumsImpl.java (hunk 2 above)
Comment (weidongxu-microsoft) [reply partial-update-tests/generated/src/main/java/fixtures/bodystring/implementation/EnumsImpl.java:125]: "Accept" is one synthesized by m4, not one in swagger.

But yes, Haoling, please verify the order is same with vanilla.

File: javagen/src/main/java/com/azure/autorest/mapper/ClientParameterMapper.java [hunk 3]
```diff
@@ -47,11 +44,20 @@ public ClientMethodParameter map(Parameter parameter) {
...
+                if(wireType instanceof EnumType) {
+                    wireType = ClassType.String;
+                }
+                if(wireType instanceof IterableType && ((IterableType) wireType).getElementType() instanceof EnumType) {
+                    wireType = new IterableType(ClassType.String);
+                }
+                if(wireType instanceof ListType && ((ListType) wireType).getElementType() instanceof EnumType) {
+                    wireType = new ListType(ClassType.String);
+                }
             }
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/mapper/ClientParameterMapper.java:61]: Refactor these same code (3 places?) to e.g. SchemaUtil.

File: partial-update-tests/generated/src/main/java/fixtures/bodystring/implementation/EnumsImpl.java (hunk 2 above)
Comment (haolingdong-msft) [reply partial-update-tests/generated/src/main/java/fixtures/bodystring/implementation/EnumsImpl.java:125]: I checked vanilla, it also put 'Accept' header parameter after body parameter.

https://github.com/Azure/autorest.java/blob/main/vanilla-tests/src/main/java/fixtures/bodystring/implementation/EnumsImpl.java#L59-L66

File: javagen/src/main/java/com/azure/autorest/mapper/ClientParameterMapper.java [hunk 4]
```diff
@@ -7,10 +7,7 @@
 import com.azure.autorest.extension.base.model.codemodel.Parameter;
 import com.azure.autorest.extension.base.model.codemodel.RequestParameterLocation;
 import com.azure.autorest.extension.base.plugin.JavaSettings;
-import com.azure.autorest.model.clientmodel.ClassType;
-import com.azure.autorest.model.clientmodel.ClientMethodParameter;
-import com.azure.autorest.model.clientmodel.IType;
-import com.azure.autorest.model.clientmodel.PrimitiveType;
+import com.azure.autorest.model.clientmodel.*;
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/mapper/ClientParameterMapper.java:10]: don't do it.

File: javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplateBase.java [hunk 5]
```diff
@@ -108,12 +108,15 @@ private static void optionalParametersJavadoc(String title, List<ProxyMethodPara
         commentBlock.line(String.format("    <caption>%s</caption>", title));
         commentBlock.line("    <tr><th>Name</th><th>Type</th><th>Required</th><th>Description</th></tr>");
         for (ProxyMethodParameter parameter : parameters) {
-            commentBlock.line(String.format(
-                    "    <tr><td>%s</td><td>%s</td><td>%s</td><td>%s</td></tr>",
-                    parameter.getRequestParameterName(),
-                    CodeNamer.escapeXmlComment(parameter.getClientType().toString()),
-                    parameter.getIsRequired() ? "Yes" : "No",
-                    parameterDescriptionOrDefault(parameter)));
+            if (parameter.getIsRequired() == false) {
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplateBase.java:111]: `!parameter.getIsRequired()`

File: generate [hunk 6]
```diff
@@ -92,6 +92,7 @@ autorest $PROTOCOL_ARGUMENTS --input-file=$SWAGGER_PATH/dpg_update1.json --names
 autorest $PROTOCOL_ARGUMENTS --input-file=$SWAGGER_PATH/parameterized-endpoint.json --namespace=fixtures.parameterizedendpoint --generate-send-request-method
 autorest $PROTOCOL_ARGUMENTS --input-file=$SWAGGER_PATH/constants.json --namespace=fixtures.constants
 autorest $PROTOCOL_ARGUMENTS --input-file=vanilla-tests/swagger/special-header.json --namespace=fixtures.specialheader
+autorest %PROTOCOL_ARGUMENTS% --input-file=protocol-tests/swagger/required-header-query.json --namespace=fixtures.requiredheaderquery
```
Comment (weidongxu-microsoft) [inline generate:95]: wrong

**Comment by weidongxu-microsoft** [comment]
Also run `python protocol-sdk-integration-tests/test.py --all` manually.

It regens the few DPG SDK we released (and a few that not included).

File: preprocessor/readme.md [hunk 7]
```diff
@@ -55,6 +55,7 @@ modelerfour:
 
 require:
   - $(this-folder)data-plane.md
+
```
Comment (weidongxu-microsoft) [inline preprocessor/readme.md:58]: revert the empty change

File: protocol-tests/src/main/java/fixtures/requiredheaderquery/DpgRequiredHeaderQueryClient.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+     * @throws ResourceModifiedException thrown if the request is rejected by server on status code 409.
+     * @return required Query Parameters along with {@link Response}.
+     */
+    @Generated
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Response<BinaryData> getRequiredQueryParamWithResponse(
+            int parameterInt,
+            boolean parameterBoolean,
+            List<String> parameterCsvStringArray,
+            List<String> parameterCsvIntArray,
```
Comment (weidongxu-microsoft) [inline protocol-tests/src/main/java/fixtures/requiredheaderquery/DpgRequiredHeaderQueryClient.java:73]: old code seems wrong, this should be `List<Integer>`

File: javagen/src/main/java/com/azure/autorest/mapper/ProxyParameterMapper.java [hunk 9]
```diff
@@ -62,11 +63,12 @@ public ProxyMethodParameter map(Parameter parameter) {
...
-        if (settings.isDataPlaneClient() && !(clientType instanceof PrimitiveType)) {
+
+        if (settings.isDataPlaneClient()) {
             if (parameterRequestLocation == RequestParameterLocation.BODY /*&& parameterRequestLocation != RequestParameterLocation.FormData*/) {
                 clientType = ClassType.BinaryData;
-            } else {
-                clientType = ClassType.String;
+            } else if(!(clientType instanceof PrimitiveType)){
+                clientType = SchemaUtil.mapDataPlaneNonPrimitiveParameterType(clientType);
             }
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/mapper/ProxyParameterMapper.java:72]: Why not put the `body` part into the utility method as well?

Just call the method `removeModelFromParameter`

**Review by weidongxu-microsoft — APPROVED** [review]

**Comment by weidongxu-microsoft** [comment]
Also run python protocol-sdk-integration-tests/test.py --all manually, before final PR.

**Pull request merged**